### PR TITLE
[SPARK-38062][CORE] Avoid resolving placeholder hostname for FallbackStorage in BlockManagerDecommissioner

### PR DESCRIPTION
**Pull request by xkrogen**
### What changes were proposed in this pull request?
This updates `BlockManagerDecommissioner` to avoid treating "remote", the placeholder hostname used by `FALLBACK_BLOCK_MANAGER_ID`, as a valid hostname and attempting to perform a network transfer to it. If the `peer` it encounters matches the fallback block manager ID, it now goes directly to accessing `fallbackStorage`, instead of first attempting to treat it like a valid block manager ID.

In addition, this reverts the changes from SPARK-37318, which should no longer be necessary now that the underlying issue is resolved.

### Why are the changes needed?
See SPARK-38062 for a much more detailed explanation. The gist of it is that:
- Attempting to resolve "remote" can behave unexpectedly in some DNS environments. This can cause failures of the `FallbackStorageSuite` tests, but also could potentially cause issues in a production deployment.
- SPARK-37318 "fixes" the tests by skipping them if such a DNS environment is detected, but this has the obvious drawback of disabling the tests, and doesn't address the problem for production environments.
- Even if resolving "remote" does quickly fail, as the current code expects, it is semantically wrong -- we should not treat this placeholder as a valid hostname.

### Does this PR introduce _any_ user-facing change?
`FallbackStorage` may be resolved slightly quicker, as it removes an unnecessary lookup step, but it should be negligible in most environments. No other user-facing changes.

### How was this patch tested?
The DNS environment in which unit tests are run in an automated fashion at my company means that we experience an issue very similar to what is described in SPARK-37318. Without this patch, tests in `FallbackStorageSuite` consistently fail, exceeding their timeouts. With this patch, the tests consistently (and quickly!) succeed.

### CODE REVIEW

**Comment by xkrogen** [comment]
cc @viirya @dongjoon-hyun who worked on SPARK-37318
cc @holdenk @attilapiros who worked on this feature initially
cc @Ngone51 @cloud-fan who I see worked on BlockManagerDecommissioner fairly recently

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me. I'll take a closer look Tonight.

**Comment by xkrogen** [comment]
@dongjoon-hyun or others, any thoughts on this PR?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala [hunk 1]
```diff
@@ -131,7 +135,10 @@ private[storage] class BlockManagerDecommissioner(
                 // driver a no longer referenced RDD with shuffle files.
                 if (bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).size < blocks.size) {
                   logWarning(s"Skipping block $shuffleBlockInfo, block deleted.")
-                } else if (fallbackStorage.isDefined) {
+                } else if (fallbackStorage.isDefined
+                  // Confirm peer is not the fallback BM ID because then fallbackStorage would
+                  // already have been used in the try-block above and there's no point trying again
+                  && peer != FallbackStorage.FALLBACK_BLOCK_MANAGER_ID) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:141]: indentation?

File: core/src/test/scala/org/apache/spark/storage/FallbackStorageSuite.scala [hunk 2]
```diff
@@ -42,13 +41,6 @@ import org.apache.spark.util.Utils.tryWithResource
 class FallbackStorageSuite extends SparkFunSuite with LocalSparkContext {
 
   def getSparkConf(initialExecutor: Int = 1, minExecutor: Int = 1): SparkConf = {
-    // Some DNS always replies for all hostnames including unknown host names
-    try {
-      InetAddress.getByName(FallbackStorage.FALLBACK_BLOCK_MANAGER_ID.host)
-      assume(false)
-    } catch {
-      case _: UnknownHostException =>
-    }
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/storage/FallbackStorageSuite.scala:51]: I didn't verify this part yet.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please enable GitHub Action on your fork in order to verify the PR passes all UTs.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala (hunk 1 above)
Comment (xkrogen) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:141]: It looks correct to me. I also just tried running `./dev/scalafmt` and it used this indentation. Do you have another opinion on what it should be?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:141]: Ya, I know that is a common mistake and we are mixing them frequently. @xkrogen .
We fallback to [Java convention](https://www.oracle.com/java/technologies/javase/codeconventions-indentation.html#248) if there is no rule in Scala style.

Your style falls into `//DON'T USE THIS INDENTATION` example exactly in Java convention. In short, Line 141 should not have the same indentation of line 140 for readability.

Comment (xkrogen) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala]: Got it, thanks @dongjoon-hyun ! Fixed.

**Comment by xkrogen** [comment]
@dongjoon-hyun GA has long been enabled on my fork, not sure what the issue was. Anyway, there is now a clean test run associated with the PR.

I accidentally introduced some changes in `pom.xml` to this PR temporarily, and though they are now reverted, the PR is still tagged with `BUILD`. Can you help me to remove the tag?

**Comment by xkrogen** [comment]
@dongjoon-hyun or @viirya , any new thoughts here?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala [hunk 3]
```diff
@@ -109,17 +109,21 @@ private[storage] class BlockManagerDecommissioner(
...
-                bm.blockTransferService.uploadBlockSync(
-                  peer.host,
-                  peer.port,
-                  peer.executorId,
-                  blockId,
-                  buffer,
-                  StorageLevel.DISK_ONLY,
-                  null) // class tag, we don't need for shuffle
-                logDebug(s"Migrated sub-block $blockId")
+              if (fallbackStorage.isDefined && peer == FallbackStorage.FALLBACK_BLOCK_MANAGER_ID) {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:112]: is the fallback host name case insensitive?

**Review by cloud-fan — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala (hunk 3 above)
Comment (xkrogen) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:112]: It is only set by this constant (e.g. as shared below), so keeping it case-sensitive should be fine.
https://github.com/apache/spark/blob/3022fd4ccfed676d4ba194afbfde2dd5ec1d348f/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L1684

There is also one area of precedence for comparing against this in a case-sensitive way:
https://github.com/apache/spark/blob/c2df895a8cbdfcbffc320e837ad1826933685912/core/src/main/scala/org/apache/spark/status/AppStatusStore.scala#L93

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by xkrogen** [comment]
Thanks @cloud-fan !

**Pull request closed without merging**